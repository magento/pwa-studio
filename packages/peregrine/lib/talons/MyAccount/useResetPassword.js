import { useState, useMemo, useCallback } from 'react';
import { useLocation } from 'react-router-dom';
import { useMutation } from '@apollo/client';

import { useGoogleReCaptcha } from '@magento/peregrine/lib/hooks/useGoogleReCaptcha';

/**
 * Returns props necessary to render a ResetPassword form.
 *
 * @param {function} props.mutations - mutation to call when the user submits the new password.
 *
 * @returns {ResetPasswordProps} - GraphQL mutations for the reset password form.
 *
 * @example <caption>Importing into your project</caption>
 * import { useResetPassword } from '@magento/peregrine/lib/talons/MyAccount/useResetPassword.js';
 */
export const useResetPassword = props => {
    const { mutations } = props;

    const [hasCompleted, setHasCompleted] = useState(false);
    const location = useLocation();
    const [
        resetPassword,
        { error: resetPasswordErrors, loading }
    ] = useMutation(mutations.resetPasswordMutation);

    const {
        recaptchaLoading,
        generateReCaptchaData,
        recaptchaWidgetProps
    } = useGoogleReCaptcha({
        currentForm: 'CUSTOMER_FORGOT_PASSWORD',
        formAction: 'resetPassword'
    });

    const searchParams = useMemo(() => new URLSearchParams(location.search), [
        location
    ]);
    const token = searchParams.get('token');

    const handleSubmit = useCallback(
        async ({ email, newPassword }) => {
            try {
                if (email && token && newPassword) {
                    const reCaptchaData = await generateReCaptchaData();

                    await resetPassword({
                        variables: { email, token, newPassword },
                        ...reCaptchaData
                    });

                    setHasCompleted(true);
                }
            } catch (err) {
                // Error is logged by apollo link - no need to double log.

                setHasCompleted(false);
            }
        },
        [generateReCaptchaData, resetPassword, token]
    );

    return {
        formErrors: [resetPasswordErrors],
        handleSubmit,
        hasCompleted,
        loading: loading || recaptchaLoading,
        token,
        recaptchaWidgetProps
    };
};

/** JSDocs type definitions */

/**
 * GraphQL mutations for the reset password form.
 * This is a type used by the {@link useResetPassword} talon.
 *
 * @typedef {Object} ResetPasswordMutations
 *
 * @property {GraphQLAST} resetPasswordMutation mutation for resetting password
 *
 * @see [resetPassword.gql.js]{@link https://github.com/magento/pwa-studio/blob/develop/packages/venia-ui/lib/components/MyAccount/ResetPassword/resetPassword.gql.js}
 * for the query used in Venia
 */

/**
 * Object type returned by the {@link useResetPassword} talon.
 * It provides props data to use when rendering the reset password form component.
 *
 * @typedef {Object} ResetPasswordProps
 *
 * @property {Array} formErrors A list of form errors
 * @property {Function} handleSubmit Callback function to handle form submission
 * @property {Boolean} hasCompleted True if password reset mutation has completed. False otherwise
 * @property {Boolean} loading True if form awaits events. False otherwise
 * @property {String} token token needed for password reset, will be sent in the mutation
 * @property {Object} recaptchaWidgetProps Props for the GoogleReCaptcha component
 */
