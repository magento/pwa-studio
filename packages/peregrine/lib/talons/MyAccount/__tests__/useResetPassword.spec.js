import React from 'react';
import { useLocation } from 'react-router-dom';
import { useMutation } from '@apollo/client';
import { act } from 'react-test-renderer';

import createTestInstance from '@magento/peregrine/lib/util/createTestInstance';

import { useResetPassword } from '../useResetPassword';

jest.mock('@apollo/client', () => ({
    useMutation: jest
        .fn()
        .mockReturnValue([jest.fn(), { error: null, loading: false }])
}));
jest.mock('react-router-dom', () => ({
    useLocation: jest.fn().mockReturnValue({
        search: '?token=eUokxamL1kiElLDjo6AQHYFO4XlK3'
    })
}));

jest.mock('@magento/peregrine/lib/hooks/useGoogleReCaptcha', () => ({
    useGoogleReCaptcha: jest.fn().mockReturnValue({
        recaptchaLoading: false,
        generateReCaptchaData: jest.fn(() => {}),
        recaptchaWidgetProps: {}
    })
}));

const Component = props => {
    const talonProps = useResetPassword(props);

    return <i talonProps={talonProps} />;
};

const getTalonProps = props => {
    const tree = createTestInstance(<Component {...props} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;

    const update = newProps => {
        act(() => {
            tree.update(<Component {...{ ...props, ...newProps }} />);
        });

        return root.findByType('i').props.talonProps;
    };

    return { talonProps, tree, update };
};

test('should render properly', () => {
    const { talonProps } = getTalonProps({
        mutations: {
            resetPasswordMutation: 'resetPasswordMutation'
        }
    });

    expect(talonProps).toMatchSnapshot();
});

test('should set hasCompleted to true if submission is successful', async () => {
    const resetPassword = jest.fn().mockResolvedValueOnce({});
    useMutation.mockReturnValueOnce([
        resetPassword,
        {
            loading: false,
            error: null
        }
    ]);
    const { talonProps, update } = getTalonProps({
        mutations: {
            resetPasswordMutation: 'resetPasswordMutation'
        }
    });

    await talonProps.handleSubmit({
        email: 'gooseton@adobe.com',
        newPassword: 'NEW_PASSWORD'
    });
    const newTalonProps = update();

    expect(newTalonProps.hasCompleted).toBeTruthy();
    expect(resetPassword).toHaveBeenCalledWith({
        variables: {
            email: 'gooseton@adobe.com',
            token: 'eUokxamL1kiElLDjo6AQHYFO4XlK3',
            newPassword: 'NEW_PASSWORD'
        }
    });
});

test('should set hasCompleted to false if submission is not successful', async () => {
    const resetPassword = jest.fn().mockRejectedValueOnce();
    useMutation.mockReturnValueOnce([
        resetPassword,
        {
            loading: false,
            error: null
        }
    ]);
    const { talonProps, update } = getTalonProps({
        mutations: {
            resetPasswordMutation: 'resetPasswordMutation'
        }
    });

    await talonProps.handleSubmit({
        email: 'gooseton@adobe.com',
        newPassword: 'NEW_PASSWORD'
    });
    const newTalonProps = update();

    expect(newTalonProps.hasCompleted).toBeFalsy();
    expect(resetPassword).toHaveBeenCalled();
});

test('should not run mutation if token is missing', async () => {
    const resetPassword = jest.fn().mockRejectedValueOnce();
    useLocation.mockReturnValueOnce({
        search: ''
    });
    useMutation.mockReturnValueOnce([
        resetPassword,
        {
            loading: false,
            error: null
        }
    ]);
    const { talonProps, update } = getTalonProps({
        mutations: {
            resetPasswordMutation: 'resetPasswordMutation'
        }
    });

    await talonProps.handleSubmit({
        email: 'gooseton@adobe.com',
        newPassword: 'NEW_PASSWORD'
    });

    update();

    expect(resetPassword).not.toHaveBeenCalled();
});
