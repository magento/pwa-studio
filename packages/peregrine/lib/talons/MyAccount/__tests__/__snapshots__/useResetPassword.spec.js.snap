// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render properly 1`] = `
Object {
  "formErrors": Array [
    null,
  ],
  "handleSubmit": [Function],
  "hasCompleted": false,
  "loading": false,
  "recaptchaWidgetProps": Object {},
  "token": "eUokxamL1kiElLDjo6AQHYFO4XlK3",
}
`;
